### PR TITLE
Migrate some features from Scribble to solc-typed-ast

### PR DESCRIPTION
**Pull request by cd1m0**
This PR adds the following helper functions:
- [x] `replaceNode(oldNode, newNode)` replaces a given node in the tree with another
- [x] `isSane(units, ctx)` checks whether a given AST tree is "sane". This involves making sure that all reachable nodes are part of the same context, that parent/child relationships are correctly set, and certain property pairs are related. This is useful to check that a re-written, or manually constructed AST tree looks correct

We also add a new integration test for `replaceNode()` that checks that we can replace most nodes in the samples we have, and produce a sane tree.

### CODE REVIEW

**Comment by cd1m0** [comment]
It seems that my line coverage is unsuffucient - 94.9% for a minimum of 95%. Also it dropped from 96% last time. @blitz-1306  is there some place where I can find the line-by-line coverage?

**Comment by blitz-1306** [comment]
The `.npmrc` file is blocking upload, by exiting with error before upload begins. We can reconfigure it in such way that codecov would result an error instead of nyc/mocha, so we still would get report. My bet would be that `isSane()` / `replaceNode()` are never invoked in the tests, or they are not sufficiently covered.

You can also look though the local coverage report by running tests and then browsing `<solc-typed-ast repo path>/coverage/index.html`.

File: src/ast/sanity.ts [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+ */
+export function isSane(unit: SourceUnit, ctx: ASTContext): boolean {
+    try {
+        checkSane(unit, ctx);
+    } catch (e) {
+        if (e instanceof InsaneASTError) {
+            console.error(e);
+
+            return false;
+        }
```
Comment (blitz-1306) [inline src/ast/sanity.ts:617]: Do we intentionally silent the `InsaneASTError`? Or is it for debugging purposes?

**Review by blitz-1306 — COMMENTED** [review]
Thanks for the contribution. A minor question.

File: src/ast/sanity.ts (hunk 1 above)
Comment (cd1m0) [reply src/ast/sanity.ts:617]: Yeah - that error is just for debugging purposes - to communicate to `isSane` (or any other consumer) what exactly was the problem. If you want to do something else other than silence then you call `checkSane` directly

**Review by blitz-1306 — DISMISSED** [review]

**Comment by blitz-1306** [comment]
@cd1m0 I did some tweaks to the  code:
- `replaceNode()` and `checkDirectChildren()` can now process properties of simple `Map` type, that have `ASTNodes` as values. This covers situation with `FunctionCallOptions` and some other nodes, however we still have and edge-case with `ImportDirective`.
- `checkSane()` now properly processed nodes with `documentation: string | StructuredDocumentation | undefined`.

Coverage is now on acceptable level, however we can do a bit more passes over the codebase to clean it up.

**Comment by blitz-1306** [comment]
@cd1m0 I removed the part of the `replaceNode()`, that was responsible for getter/setter lookup in prototype chain. It is never hit due to the only candidate at the moment is `ContractDefinition` with `documentation` setter. However, even this setter is not hit as the field `ownChildren` is handled by `Object.getOwnPropertyDescriptors()`. So children are patched without any necessity to interact with setters. This also bumbed coverage a bit.

**Review by blitz-1306 — APPROVED** [review]
Looks good to merge now.

**Pull request merged**